### PR TITLE
Fix complex date restriction regression and various others

### LINKED ISSUE #824: NULL comparison
`tab & {'f': None}` should translate to `tab & "f IS NULL"`. Currently, users must use this string comparison.

### LINKED ISSUE #838: Request for a `schema.list_tables()`
Similar to `dj.list_schemas()`, would be great if we can list all tables under a particular schema that are topologically sorted - with `schema.list_tables()`

### LINKED ISSUE #882: dj.Diagram does not work for Part tables for 0.13.dev5
### Description
dj.Diagram does not show part tables as Class names, but show them as names on the database side.

### Reproducibility
Example:
image

### LINKED ISSUE #892: Complex string restrictor does not get parsed correctly in 0.13.0
### Description
When there is complex expression in restrictor string, the query does not get parsed correctly.

### Reproducibility
Include:
- OS: MACOS and Linux
- Python Version 3.8 & 3.7
- MySQL Deployment Strategy: local-native and remote
- DataJoint Version: 0.13.0

Code to reproduce:
```
schema = dj.schema('test')
@schema
class Session(dj.Manual):
    definition = """
    session_date : varchar(8)
    """

Session & 'session_date between curdate() - interval 30 day and curdate()'
```
gives error:
```
---------------------------------------------------------------------------
DataJointError                            Traceback (most recent call last)
<ipython-input-9-aff0f5e74045> in <module>
----> 1 Session & 'session_date between curdate() - interval 30 day and curdate()'

~/anaconda3/lib/python3.7/site-packages/datajoint/user_tables.py in __and__(cls, arg)
     31 
     32     def __and__(cls, arg):
---> 33         return cls() & arg
     34 
     35     def __xor__(cls, arg):

~/anaconda3/lib/python3.7/site-packages/datajoint/expression.py in __and__(self, restriction)
    198         See QueryExpression.restrict for more detail.
    199         """
--> 200         return self.restrict(restriction)
    201 
    202     def __xor__(self, restriction):

~/anaconda3/lib/python3.7/site-packages/datajoint/expression.py in restrict(self, restriction)
    173         try:
    174             raise DataJointError("Attribute `%s` is not found in query." % next(
--> 175                 attr for attr in attributes if attr not in self.heading.names))
    176         except StopIteration:
    177             pass  # all ok

DataJointError: Attribute `day` is not found in query.
```
This used to work in dj 0.12.

### LINKED ISSUE #898: Unexpected Join product in 0.13.0
## Bug Report

Join product is unexpected for dj 0.13.0, behavior was correct in dj 0.12.9

### Reproducibility
- OS (MACOS)
- Python Version 3.7.3
- MySQL Version 10.4.13-MariaDB 
- MySQL Deployment Strategy (local-native)
- DataJoint Version 0.13.0

Code to reproduce the error:
```
import datajoint as dj
schema = dj.schema('test')

@schema
class Subject(dj.Manual):
    definition = """
    subject_id: varchar(32)
    ---
    dob : date
    sex : enum('M', 'F', 'U')
    """

@schema
class Session(dj.Manual):
    definition = """
    -> Subject
    session_start_time: datetime
    ---
    session_dir=''  : varchar(32)
    """

@schema
class SessionStatus(dj.Manual):
    definition = """
    -> Session
    ---
    status: enum('in_training', 'trained_1a', 'trained_1b', 'ready4ephys')
    """

@schema
class SessionDate(dj.Manual):
    definition = """
    -> Subject
    session_date:  date
    """

Subject.insert([
    ('mouse1', '2020-09-01', 'M'),
    ('mouse2', '2020-03-19', 'F'),
    ('mouse3', '2020-08-23', 'F')
])

Session.insert([
    ('mouse1', '2020-12-01 12:32:34', ''),
    ('mouse1', '2020-12-02 12:32:34', ''),
    ('mouse1', '2020-12-03 12:32:34', ''),
    ('mouse1', '2020-12-04 12:32:34', '')
])

SessionStatus.insert([
    ('mouse1', '2020-12-01 12:32:34', 'in_training'),
    ('mouse1', '2020-12-02 12:32:34', 'trained_1a'),
    ('mouse1', '2020-12-03 12:32:34', 'trained_1b'),
    ('mouse1', '2020-12-04 12:32:34', 'ready4ephys'),
])

SessionDate.insert([
    ('mouse1', '2020-12-01'),
    ('mouse1', '2020-12-02'),
    ('mouse1', '2020-12-03'),
    ('mouse1', '2020-12-04')
])

subjects = Subject.aggr(
    SessionStatus & 'status="trained_1a" or status="trained_1b"', 
    date_trained='min(date(session_start_time))')
```

After the above commands, subjects is like this:
image
 
SessionDate is like this:
image

In 0.12.9, their join product is correct:
image

In 0.13.0, their join product is not correct:
image
It said total is 4 somehow.

### LINKED ISSUE #899: Unknown column error in restriction on a join product for dj 0.13.0
## Bug Report
Related to #898 but a little different, when I do a restriction on a join product, there is a `Unknown column error` happens. This is only seen in 0.13.0, it was fine for 0.12.9.

### Reproducibility
Same setting as #898, with the same code base:
```
import datajoint as dj
schema = dj.schema('test')

@schema
class Subject(dj.Manual):
    definition = """
    subject_id: varchar(32)
    ---
    dob : date
    sex : enum('M', 'F', 'U')
    """

@schema
class Session(dj.Manual):
    definition = """
    -> Subject
    session_start_time: datetime
    ---
    session_dir=''  : varchar(32)
    """

@schema
class SessionStatus(dj.Manual):
    definition = """
    -> Session
    ---
    status: enum('in_training', 'trained_1a', 'trained_1b', 'ready4ephys')
    """

@schema
class SessionDate(dj.Manual):
    definition = """
    -> Subject
    session_date:  date
    """

Subject.insert([
    ('mouse1', '2020-09-01', 'M'),
    ('mouse2', '2020-03-19', 'F'),
    ('mouse3', '2020-08-23', 'F')
])

Session.insert([
    ('mouse1', '2020-12-01 12:32:34', ''),
    ('mouse1', '2020-12-02 12:32:34', ''),
    ('mouse1', '2020-12-03 12:32:34', ''),
    ('mouse1', '2020-12-04 12:32:34', '')
])

SessionStatus.insert([
    ('mouse1', '2020-12-01 12:32:34', 'in_training'),
    ('mouse1', '2020-12-02 12:32:34', 'trained_1a'),
    ('mouse1', '2020-12-03 12:32:34', 'trained_1b'),
    ('mouse1', '2020-12-04 12:32:34', 'ready4ephys'),
])

SessionDate.insert([
    ('mouse1', '2020-12-01'),
    ('mouse1', '2020-12-02'),
    ('mouse1', '2020-12-03'),
    ('mouse1', '2020-12-04')
])

subj_query = Subject.aggr(
    Session * SessionStatus & 'status="trained_1a" or status="trained_1b"', date_trained='min(date(session_start_time))')
(SessionDate * (subj_query & 'date_trained<"2020-12-21"')) & 'session_date<"date_trained"'
```
gives an error:

```
---------------------------------------------------------------------------
UnknownAttributeError                     Traceback (most recent call last)
~/anaconda3/lib/python3.7/site-packages/IPython/core/formatters.py in __call__(self, obj)
    700                 type_pprinters=self.type_printers,
    701                 deferred_pprinters=self.deferred_printers)
--> 702             printer.pretty(obj)
    703             printer.flush()
    704             return stream.getvalue()

~/anaconda3/lib/python3.7/site-packages/IPython/lib/pretty.py in pretty(self, obj)
    400                         if cls is not object \
    401                                 and callable(cls.__dict__.get('__repr__')):
--> 402                             return _repr_pprint(obj, self, cycle)
    403 
    404             return _default_pprint(obj, self, cycle)

~/anaconda3/lib/python3.7/site-packages/IPython/lib/pretty.py in _repr_pprint(obj, p, cycle)
    695     """A pprint that just redirects to the normal repr function."""
    696     # Find newlines and replace them with p.break_()
--> 697     output = repr(obj)
    698     for idx,output_line in enumerate(output.splitlines()):
    699         if idx:

~/anaconda3/lib/python3.7/site-packages/datajoint/expression.py in __repr__(self)
    524         :rtype: str
    525         """
--> 526         return super().__repr__() if config['loglevel'].lower() == 'debug' else self.preview()
    527 
    528     def preview(self, limit=None, width=None):

~/anaconda3/lib/python3.7/site-packages/datajoint/expression.py in preview(self, limit, width)
    528     def preview(self, limit=None, width=None):
    529         """ :return: a string of preview of the contents of the query. """
--> 530         return preview(self, limit, width)
    531 
    532     def _repr_html_(self):

~/anaconda3/lib/python3.7/site-packages/datajoint/preview.py in preview(query_expression, limit, width)
     11     if width is None:
     12         width = config['display.width']
---> 13     tuples = rel.fetch(limit=limit + 1, format="array")
     14     has_more = len(tuples) > limit
     15     tuples = tuples[:limit]

~/anaconda3/lib/python3.7/site-packages/datajoint/fetch.py in __call__(self, offset, limit, order_by, format, as_dict, squeeze, download_path, *attrs)
    191         else:  # fetch all attributes as a numpy.record_array or pandas.DataFrame
    192             cur = self._expression.cursor(
--> 193                 as_dict=as_dict, limit=limit, offset=offset, order_by=order_by)
    194             heading = self._expression.heading
    195             if as_dict:

~/anaconda3/lib/python3.7/site-packages/datajoint/expression.py in cursor(self, offset, limit, order_by, as_dict)
    514             sql += ' LIMIT %d' % limit + (' OFFSET %d' % offset if offset else "")
    515         logger.debug(sql)
--> 516         return self.connection.query(sql, as_dict=as_dict)
    517 
    518     def __repr__(self):

~/anaconda3/lib/python3.7/site-packages/datajoint/connection.py in query(self, query, args, as_dict, suppress_warnings, reconnect)
    300         cursor = self._conn.cursor(cursor=cursor_class)
    301         try:
--> 302             self._execute_query(cursor, query, args, suppress_warnings)
    303         except errors.LostConnectionError:
    304             if not reconnect:

~/anaconda3/lib/python3.7/site-packages/datajoint/connection.py in _execute_query(cursor, query, args, suppress_warnings)
    266                 cursor.execute(query, args)
    267         except client.err.Error as err:
--> 268             raise translate_query_error(err, query)
    269 
    270     def query(self, query, args=(), *, as_dict=False, suppress_warnings=True, reconnect=None):

UnknownAttributeError: Unknown column 'date_trained' in 'where clause'

---------------------------------------------------------------------------
UnknownAttributeError                     Traceback (most recent call last)
~/anaconda3/lib/python3.7/site-packages/IPython/core/formatters.py in __call__(self, obj)
    343             method = get_real_method(obj, self.print_method)
    344             if method is not None:
--> 345                 return method()
    346             return None
    347         else:

~/anaconda3/lib/python3.7/site-packages/datajoint/expression.py in _repr_html_(self)
    532     def _repr_html_(self):
    533         """ :return: HTML to display table in Jupyter notebook. """
--> 534         return repr_html(self)
    535 
    536 

~/anaconda3/lib/python3.7/site-packages/datajoint/preview.py in repr_html(query_expression)
     32     rel = query_expression.proj(*heading.non_blobs)
     33     info = heading.table_status
---> 34     tuples = rel.fetch(limit=config['display.limit'] + 1, format='array')
     35     has_more = len(tuples) > config['display.limit']
     36     tuples = tuples[0:config['display.limit']]

~/anaconda3/lib/python3.7/site-packages/datajoint/fetch.py in __call__(self, offset, limit, order_by, format, as_dict, squeeze, download_path, *attrs)
    191         else:  # fetch all attributes as a numpy.record_array or pandas.DataFrame
    192             cur = self._expression.cursor(
--> 193                 as_dict=as_dict, limit=limit, offset=offset, order_by=order_by)
    194             heading = self._expression.heading
    195             if as_dict:

~/anaconda3/lib/python3.7/site-packages/datajoint/expression.py in cursor(self, offset, limit, order_by, as_dict)
    514             sql += ' LIMIT %d' % limit + (' OFFSET %d' % offset if offset else "")
    515         logger.debug(sql)
--> 516         return self.connection.query(sql, as_dict=as_dict)
    517 
    518     def __repr__(self):

~/anaconda3/lib/python3.7/site-packages/datajoint/connection.py in query(self, query, args, as_dict, suppress_warnings, reconnect)
    300         cursor = self._conn.cursor(cursor=cursor_class)
    301         try:
--> 302             self._execute_query(cursor, query, args, suppress_warnings)
    303         except errors.LostConnectionError:
    304             if not reconnect:

~/anaconda3/lib/python3.7/site-packages/datajoint/connection.py in _execute_query(cursor, query, args, suppress_warnings)
    266                 cursor.execute(query, args)
    267         except client.err.Error as err:
--> 268             raise translate_query_error(err, query)
    269 
    270     def query(self, query, args=(), *, as_dict=False, suppress_warnings=True, reconnect=None):

UnknownAttributeError: Unknown column 'date_trained' in 'where clause'
```

### PR DESCRIPTION
**Pull request by guzman-raphael**
This PR addresses:
- [x] Fix #824
- [x] Fix #838
- [x] Fix #882
- [x] Fix #892
- [x] Fix #898
- [x] Fix #899

### CODE REVIEW

**Review by ixcat — APPROVED** [review]

File: datajoint/diagram.py [hunk 1]
```diff
@@ -219,24 +219,31 @@ def _make_graph(self):
             """
             Make the self.graph - a graph object ready for drawing
             """
-            # mark "distinguished" tables, i.e. those that introduce new primary key attributes
+            # mark "distinguished" tables i.e. ones introduce new primary key attributes
```
Comment (dimitri-yatsenko) [inline datajoint/diagram.py:222]: The wording before was more correct.

**Review by dimitri-yatsenko — APPROVED** [review]
great improvements! Just minor comments.

File: tests/test_relational_operand.py [hunk 2]
```diff
@@ -459,3 +460,23 @@ def test_permissive_join_basic():
     def test_permissive_restriction_basic():
         """Verify join compatibility check is skipped for restriction"""
         Child ^ Parent
+
+    @staticmethod
+    def test_complex_date_restriction():
+        # https://github.com/datajoint/datajoint-python/issues/892
+        """Test a complex date restriction"""
+        q = OutfitLaunch & 'day between curdate() - interval 30 day and curdate()'
```
[the remark this answers is not in the document]
Comment (guzman-raphael) [reply tests/test_relational_operand.py:539]: Addressed

File: datajoint/diagram.py (hunk 1 above)
Comment (guzman-raphael) [reply datajoint/diagram.py]: Addressed

**Review by dimitri-yatsenko — APPROVED** [review]

File: CHANGELOG.md [hunk 3]
```diff
@@ -1,21 +1,31 @@
 ## Release notes
 
+### 0.13.1 -- TBD
+* Add `None` as an alias for `NULL` in `dict` restrictions (#824) PR #893
```
Comment (dimitri-yatsenko) [inline CHANGELOG.md:4]:
```suggestion
* Add `None` as an alias for `IS NULL` comparison in `dict` restrictions (#824) PR #893
```

Comment (guzman-raphael) [reply CHANGELOG.md]: Addressed.

**Pull request merged**